src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.MessageMode.Windows.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO.Tests;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Pipes.Tests
+{
+    // Support for PipeAccessRights and setting ReadMode to Message is only supported on Windows
+    public class NamedPipeTest_MessageMode_Windows
+    {
+        private const PipeAccessRights MinimumMessageAccessRights = PipeAccessRights.ReadData | PipeAccessRights.WriteAttributes;
+
+        private static NamedPipeClientStream CreateClientStream(string pipeName, PipeOptions options) =>
+            new NamedPipeClientStream(".", pipeName, MinimumMessageAccessRights, options, Security.Principal.TokenImpersonationLevel.None, HandleInheritability.None);
+
+        [Theory]
+        [InlineData(PipeDirection.Out, PipeOptions.None)]
+        [InlineData(PipeDirection.InOut, PipeOptions.Asynchronous)]
+        public async Task Client_DetectsMessageCompleted(PipeDirection serverDirection, PipeOptions options)
+        {
+            string pipeName = PipeStreamConformanceTests.GetUniquePipeName();
+
+            using NamedPipeServerStream server = new NamedPipeServerStream(pipeName, serverDirection, 1, PipeTransmissionMode.Message, options);
+            using NamedPipeClientStream client = CreateClientStream(pipeName, options);
+
+            Task.WaitAll(server.WaitForConnectionAsync(), client.ConnectAsync());
+            client.ReadMode = PipeTransmissionMode.Message;
+
+            ValueTask serverWrite = server.WriteAsync(new byte[] { 1, 2, 3, 4, 5 });
+
+            byte[] buffer1 = new byte[2], buffer2 = new byte[2], buffer3 = new byte[2];
+            bool[] messageCompleted = new bool[3];
+
+            int bytesRead = client.Read(buffer1, 0, 2);
+            messageCompleted[0] = client.IsMessageComplete;
+
+            bytesRead += client.Read(buffer2, 0, 2);
+            messageCompleted[1] = client.IsMessageComplete;
+
+            bytesRead += client.Read(buffer3, 0, 2);
+            messageCompleted[2] = client.IsMessageComplete;
+
+            Assert.Equal(5, bytesRead);
+            Assert.Equal(new byte[] { 1, 2, 3, 4, 5, 0 }, buffer1.Concat(buffer2).Concat(buffer3));
+            Assert.Equal(new bool[] { false, false, true }, messageCompleted);
+
+            await serverWrite;
```

Comment:
I would move line 35 to here, and return the `ValueTask` directly:

```suggestion
            return server.WriteAsync(new byte[] { 1, 2, 3, 4, 5 });
```

And since you have no more awaits in the method body, you can remove the `async` modified from the method. And I think you can also change the return type of the method to `ValueTask`.

@stephentoub please yell at me if I'm giving a bad suggestion.

Reply:
I'm not sure I follow the recommendation. The test is getting both the server and client running and connected, and then it fires off an async write from the server so that the test can proceed with the client consuming the written bytes (and make asserts based on the received data). Without the `WriteAsync`, the `client.Read` calls will block until data can be received.